File: 1/2020.10.14/198.House Robber/java/solution.java
class Solution {
    public int rob(int[] nums) {
        // dp, bottom -> top
        if (nums.length == 0) {
            return 0;
        }
        int first = 0, second = 0;
        for (int num : nums) {
            int tmp = Math.max(second, first + num);
            first = second;
            second = tmp;
        }
        return second;
    }

    public int rob(int[] nums) {
        // dp, top -> bottom
        int[] memo = new int[nums.length];
        Arrays.fill(memo, -1);
        return rob(nums, nums.length - 1, memo);
    }

    private int rob(int[] nums, int n, int[] memo) {
        if (n < 0) {
            return 0;
        }
        if (memo[n] != -1) {
            return memo[n];
        }
        int res = Math.max(rob(nums, n - 2, memo) + nums[n], rob(nums, n - 1, memo));
        memo[n] = res;
        return res;
    }
}

File: 1/2020.10.14/198.House Robber/python/solution.py
import functools
class Solution:
    def rob(self, nums: List[int]) -> int:
        # dp, top -> bottom
        @functools.lru_cache(len(nums))
        def _rob(n):
            if n == 0:
                return nums[n]
            if n < 0:
                return 0
            return max(nums[n] + _rob(n-2), _rob(n-1))
        if not nums:
            return 0
        return _rob(len(nums) - 1)

    def rob(self, nums: List[int]) -> int:
        # dp, bottom -> top
        if not nums:
            return 0
        first = second = 0
        for num in nums:
            first, second = second, max(second, first + num)
        return second
